---
title: "promoters-in-R"
author: "zagor and val"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output: 
  html_document:
    toc: true # table of content true
    toc_depth: 3  # upto three depths of headings (specified by #, ## and ###)
    number_sections: true  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```




# packages

```{r}



library(Rsamtools)
library(GenomicFeatures)




```

# data



## fasta


```{r}

my.directory = file.path('..', 'input', 'PLAZA')
my.file.1 = 'sly.fasta'



fa <- open(FaFile(file.path(my.directory, my.file.1)))

str(fa)

```


## gff3/gtf to TxDb
```{r}



my.file.2 = 'annotation.all_transcripts.all_features.sly.gff3'


potato.gff3 = GenomicFeatures::makeTxDbFromGFF(file.path(my.directory, my.file.2),
                              format="gff3",
                              dataSource=NA,
                              # organism='Solanum tuberosum',
                              taxonomyId=NA,
                              circ_seqs=NULL,
                              chrominfo=NULL,
                              miRBaseBuild=NA,
                              metadata=NULL)
str(potato.gff3)



```


## IDs
```{r}

my.directory = file.path('..', 'input')
my.file.3 = 'v4IDs.txt'




geneIDs = read.delim(file = file.path(my.directory, my.file.3), 
                    header = FALSE, 
                    sep = "\n", 
                    quote = NULL,
                    dec = ".", 
                    fill = TRUE, 
                    comment.char = "@")
geneIDs = unlist(geneIDs)
length(geneIDs)
length(unique(geneIDs))

sum(grepl('DMG', geneIDs))

transcriptCoordsByGene.GRangesList = cdsBy(potato.gff3, by = "gene") 
transcriptCoordsByGene.GRangesList




```

# +/- exons

```{r}

myVec = NULL

for (i in (1:length(transcriptCoordsByGene.GRangesList))){

  if (!(i %% 1000)) print(i)
  strand = transcriptCoordsByGene.GRangesList[[i]]@strand@values
  start = transcriptCoordsByGene.GRangesList[[i]]@ranges@start
  ind = ifelse(strand == '-', which(start == max(start)), which(start == min(start)))
  transcriptCoordsByGene.GRangesList[[i]] = transcriptCoordsByGene.GRangesList[[i]][ind,]
 
  myVec = c(myVec, paste0(transcriptCoordsByGene.GRangesList[[i]]@elementMetadata[,2],
                         '|Exon-order-in-GFF:',
                         stringr::str_pad(ind, 3, pad = "0"),
                         '|',
                         strand,
                         'strand'
                         ))
  
}


```



# upstream/downstream

the most right exon is the first one

mirrored order (right to left) when (-) strain in play


getPromoterSeq is strand-aware in two ways.  For gene transcripts annotated to the minus strand:

   1) It interprets 'upstream' as 'increasing chromosomal coordinates'.
   2) It returns the reverse complement of the extracted sequence.


```{r}


upstream.width = 1000
downstream.width = 0

my.directory = file.path('..', 'input', 'PLAZA')
dna = rtracklayer::import(file.path(my.directory, my.file.1))
width(dna)
names(dna)
tmp = cbind(names(dna), width(dna))
colnames(tmp) = c('seqnames', 'chrWidth')

ranges = unlist(transcriptCoordsByGene.GRangesList)
ranges = merge(ranges, tmp, by = 'seqnames')
ranges$chrWidth = as.numeric(ranges$chrWidth)
ranges$L1000 = ranges$start-1000
ranges$R1000 = ranges$end+1000
table(ranges$R1000 >= ranges$chrWidth)
table(ranges$L1000 <= 0)
ind = which(((ranges$R1000 >= ranges$chrWidth) & (ranges$strand == '-')) | ((ranges$L1000 < 0) & (ranges$strand == '+')))
toShort = ranges[ind, ]
indd = which(transcriptCoordsByGene.GRangesList@unlistData$cds_name %in% toShort$cds_name)
length(indd)

temp = transcriptCoordsByGene.GRangesList[-indd,]

ts = cbind(abs(1000 - (toShort$R1000  - toShort$chrWidth)),
           abs(1000 + toShort$L1000))
mm = apply(ts, 1, min) - 1


promoter.seqs = getPromoterSeq (temp, #transcriptCoordsByGene.GRangesList,
                                 fa, 
                                 upstream=upstream.width, 
                                 downstream=downstream.width)




temp = transcriptCoordsByGene.GRangesList[indd,]
ind = grep(paste(temp@unlistData$cds_name, collapse = '|'), myVec)

# upstream.width = 100
promoter.seqs2 = getPromoterSeq (temp, #transcriptCoordsByGene.GRangesList,
                                 fa, 
                                 upstream=mm, 
                                 downstream=downstream.width)

names(promoter.seqs) = paste0(names(promoter.seqs), '|||', 
                              myVec[-ind], '|||', 
                              stringr::str_pad(upstream.width, 5, pad = "0"), 'nt-upstream')


names(promoter.seqs2) = paste0(names(promoter.seqs2), '|||', 
                              myVec[ind], '|||', 
                              stringr::str_pad(upstream.width, 5, pad = "0"), 'nt-upstream')

print(promoter.seqs2)
str(promoter.seqs2)
promoter.seqs@unlistData
promoter.seqs2@unlistData




```



# write fasta
```{r}




output.directory = file.path('..', 'output')
file.output.name = 'ITAGv4_1000nt_upstream-regions.fasta'



writeXStringSet(unlist(promoter.seqs), 
                file.path(output.directory, file.output.name),
                append=FALSE,
                compress=FALSE, 
                compression_level=NA, 
                format="fasta",
                width = upstream.width)



file.output.name = 'ITAGv4_short_upstream-regions.fasta'

writeXStringSet(unlist(promoter.seqs2), 
                file.path(output.directory, file.output.name),
                append=FALSE,
                compress=FALSE, 
                compression_level=NA, 
                format="fasta",
                width = upstream.width)



```



```{r}

sessionInfo()

```

